csrc/preseg_passes/move_pad.cpp

```diff
@@ -0,0 +1,530 @@
+      continue;
+    }
+
+    // unify all consumer pad of tv;
+    TensorView* tv = p->in()->as<TensorView>();
+    for (Expr* use : tv->uses()) {
+      if (use == p) {
+        continue;
+      }
+      // check if use is the same pad operation (same pad value / width e.t.c.)
+      if (isSamePadOp(use, p)) {
+        // replace consumer of use->out() with p->out()
+        ir_utils::replaceValInAllExprInputsAndFusionOutputs(
+            use->output(0), p->out());
+        fusion->removeExpr(use);
+      }
+    }
+
+    // if tv is fusion output, we need to keep tv alive, it might render
+    // propagating PadOp before tv->definition() being non-optimal.
+    if (tv->isFusionOutput()) {
+      continue;
+    }
+
+    // check for pad_dependencies to verify that 'p' can be moved before 'def'.
+    std::unordered_set<Val*> pad_inputs;
+    for (Val* val : p->inputs()) {
+      if (val == p->in() || val->isConst()) {
+        continue;
+      }
+      pad_inputs.insert(val);
+    }
+    std::unordered_set<Val*> pad_dependencies =
+        DependencyCheck::getAllDependentVals(pad_inputs);
+    auto pad_replay_check = [&pad_dependencies](Expr* expr) {
+      return std::all_of(
+          expr->inputs().begin(),
+          expr->inputs().end(),
+          [&pad_dependencies = std::as_const(pad_dependencies)](Val* val) {
+            if (!val->isA<TensorView>()) {
+              return false;
+            }
+            if (pad_dependencies.count(val) > 0) {
+              return false;
+            }
+            return true;
+          });
+    };
+
+    Expr* def = p->in()->definition();
+    Val* new_out = nullptr;
+
+    if (auto* uop = dynamic_cast<UnaryOp*>(def)) {
+      // stop propagation if current PadOp p isn't the only use of tv, since it
+      // requires tv to be live in the fusion.
+      if (tv->uses().size() != 1) {
+        continue;
+      }
+      NVF_ERROR(
+          tv->uses()[0] == p,
+          "expect existing PadOp to be the only use of its input");
+      // check if unary op type is compatible for zero pad propagation.
+      if (!zeroIsFixedPoint(uop->getUnaryOpType())) {
+        continue;
+      }
+      // check if PadOp can be replayed on input(s)
+      if (!pad_replay_check(uop)) {
+        continue;
+      }
+
+      // replay pad on input(s)
+      Val* new_pad_out = replayConcretePad(
+          uop->in()->as<TensorView>(),
+          p->value(),
+          {p->getPadWidths()},
+          TensorDomain::noReductions(
+              p->out()->as<TensorView>()->getLogicalDomain()));
+
+      new_out = ops::newValLike(new_pad_out, uop->out()->getDataType().value());
+      IrBuilder::create<UnaryOp>(uop->getUnaryOpType(), new_out, new_pad_out);
+      // insert new PadOp(s) to frontier;
+      frontier.push_back(new_pad_out->definition()->as<PadOp>());
+      simple_pad_set.insert(new_pad_out->definition()->as<PadOp>());
+    } else if (auto* bop = dynamic_cast<BinaryOp*>(def)) {
```

Comment:
I'm seeing a repeated pattern in handling unary, binary, and cat. Have you thought about combining them?